lms/static/completion/js/ViewedEvent.js

```diff
@@ -0,0 +1,158 @@
+function throttle(fn, wait) {
+  let time = Date.now();
+  function delay() {
+    const now = Date.now();
+    if (time + wait < now) {
+      time = now;
+      fn();
+    }
+  }
+  return delay;
+}
+
+export class ElementViewing {
+  constructor(el, completeAfterMs) {
+    this.el = el;
+    this.completeAfterMs = completeAfterMs;
+    this.topSeen = false;
+    this.bottomSeen = false;
+    this.seenForMs = 0;
+    this.lastSeen = undefined;
+    this.complete = false;
+  }
+
+  getBoundingRect() {
+    return this.el.getBoundingClientRect();
+  }
+
+  pauseTracking() {
+    this.lastSeen = undefined;
+  }
+
+  markTopSeen() {
+    this.topSeen = true;
+  }
+
+  markBottomSeen() {
+    this.bottomSeen = true;
+  }
+
+  hasBeenViewed() {
+    return this.topSeen && this.bottomSeen && (this.seenForMs >= this.completeAfterMs);
+  }
+
+  handleVisible(now) {
+    if (this.lastSeen) {
+      this.seenForMs += (now - this.lastSeen);
+    }
+    this.lastSeen = now;
+  }
+
+  sendEvent(callback) {
+    callback(this.el, {
+      elementHasBeenViewed: this.hasBeenViewed(),
+    });
+  }
+}
+
+
+export class ViewedEventTracker {
+  /**
+   * When the top or bottom of an element is
+   * first viewed, or the entire element is viewed for a specified amount
+   * of time, the callback is called, passing the element that was viewed,
+   * and an event object having the following field:
+   *
+   * *   hasBeenViewed (bool): true if all the conditions for being
+   *     considered "viewed" have been met.
+   */
+  constructor(elements, completeAfterMs) {
+    this.completeAfterMs = completeAfterMs;
+    this.elementViewings = new Set();
+    elements.forEach((el) => {
+      this.elementViewings.add(new ElementViewing(el, completeAfterMs));
+    });
+    this.visible = new Set();
+    this.previouslyVisible = new Set();
+    this.handlers = [];
+    this.interval = undefined;
+    this.registerDomHandlers();
+  }
+
+  /** Register a new handler to be called when an element has been viewed.  */
+  addHandler(handler) {
+    this.handlers.push(handler);
+  }
+
+  windowHeight() { // eslint-disable-line class-methods-use-this
```

Comment:
I don't think this should be a method - is there a reason not to just use `window.innerHeight` directly on line 92?

Reply:
I was planning on needing to mock it in the tests, but it turned out I didn't have to.